src/storage-types/src/read_holds.rs

```diff
@@ -41,6 +41,19 @@ impl<T: TimelyTimestamp> Debug for ReadHold<T> {
     }
 }
 
+/// Errors for manipulating read holds.
+#[derive(Error, Debug)]
+pub enum ReadHoldDowngradeError<T> {
+    /// The new frontier is not beyond the current since.
+    #[error("since violation: new frontier {frontier:?} is not beyond current since {since:?}")]
+    SinceViolation {
+        /// The frontier to downgrade to.
+        frontier: Antichain<T>,
+        /// The since of the collection.
+        since: Antichain<T>,
+    },
+}
```

Comment:
There is also a `ReadHoldError` at the bottom of this file. This one has an additional `CollectionMissing` variant that doesn't apply to downgrading. It's nice to have the returned error reflect exactly the possible error conditions, so having two errors seems like a good idea. But it might be confusing to have two errors with very similar names.

@aljoscha Should we move `ReadHoldError` somewhere else? It's not used in this module at all, and it seems to be an error that's only produce by the storage controller, so seems like it belongs in `mz-storage-client` instead?

Reply:
I thought the same, it should really be used elsewhere. Didn't want to change it as part of this PR because it addresses an actual issue!